### ISSUE TITLE
evergreen-layers outline

### ISSUE THREAD

**Issue by jeroenransijn** [segmentio/evergreen#8]
The `evergreen-layers` is a package exporting React components and layer/elevation styles.
Not entirely sure about some of the API specifics, but have a pretty good idea.

## Usage

```jsx
import {
  Pane,
  Card,
  ElevationStyles,
  BorderColors,
  LayerAppearances
}  from 'evergreen-layers'

<Card elevation={0} />
```

## Design example

<img width="1242" alt="screen shot 2017-08-30 at 8 51 11 pm" src="https://user-images.githubusercontent.com/564463/29905859-e959d488-8dc5-11e7-9654-ed84594c23ae.png">

## `<Pane />` component

Pane will simply implement `ui-box` with some opinionated properties. It is used pretty much as a `div` or `Box` element and acts as a abstraction between `Box`.

```jsx
<Pane appearance="dark" elevation={2} />
```

The styles will be implemented on a `LayerAppearances` object.

```javascript
const LayerAppearances = {
   dark: { backgroundColor: colors.neutral['800'] }
   selected: {
       backgroundColor: colors.blue['10A'],
       boxShadow: `inset 0 0 0 1px ${colors.blue['30A']}`,
       borderColor: 'selected'
   }
   // are there some obvious colors I am forgetting?
}
```

#### Selected appearance

In some cases you want to make the Card interactive, therefore having a selected appearance is often a nice to have.

<img width="250" alt="screen shot 2017-08-30 at 9 39 36 pm" src="https://user-images.githubusercontent.com/564463/29906763-f9500c80-8dcb-11e7-8247-59a220468307.png">

One way of dealing with the appearance to avoid elevation conflicts is to have the appearance applied after the elevation box shadow, this would cancel out that box shadow.

## `<Card />` component

The `Card` component directly implements the Pane component. The only difference is that it comes with opinionated border radius. Generally engineers and designers like to refer to these things as Cards.

The standard border radius will be `5px`. Maybe will make this configurable in the future.

```jsx
<Card elevation={2} />
```

## Elevation property

The `elevation` property determines what kind of box shadow will be set.

```jsx
<Pane elevation={0} />
<Pane elevation={1} />
<Pane elevation={2} />
<Pane elevation={3} />
<Pane elevation={4} />
```

I don't think there should be a default for the elevation if you don't need a box shadow.

### `hoverElevation`, and `activeElevation` property

To easily allow for interactive cards, having a `hoverElevation` and `activeElevation` is a nice addition. This will have to use the Box `css `.

### Elevation box shadows implementation

The box shadows themselves will be implemented on a `ElevationStyles` array.
The box shadow is following a technique where you use 2 box shadows, one of which is implemented with a negative spread to avoid blurry sides. Which is sadly not supported by Sketch

The color used right now is for the box shadow acting as a border is `colors.neutral['80A']`. However this has proved somewhat light in some circumstances. This might be bumped up slightly. Or instead of using a blur we'll use a spread.

```jsx
import colors from 'evergreen-colors'

const borderShadowColor = colors.neutral['80A']
const blurryShadowColor = colors.neutral['50A']

const ElevationStyles = [
  `0 0 1px ${borderShadowColor}`,
  `0 0 1px ${borderShadowColor}, 0 2px 4px -2px ${blurryShadowColor}`,
  `0 0 1px ${borderShadowColor}, 0 3px 6px -2px ${blurryShadowColor}`,
  `0 0 1px ${borderShadowColor}, 0 8px 10px -4px ${blurryShadowColor}`,
  `0 0 1px ${borderShadowColor}, 0 16px 24px -8px ${blurryShadowColor}`
]

export default ElevationStyles
```

## Border property

In the `react-ui-library` (Segment's current internal UI library) there is a shorthand for setting borders on any Box, simply by doing something like the following:

```jsx
<Pane borderBottom borderTop borderLeft borderRight /> 
```

I actually have really enjoyed using this API in the past, and I don't see a clear alternative for this.

I am thinking this should still use a actual border. I am not sure how this would work when using `elevation`. Maybe add use a box shadow at that point?

I think the border color should be using a neutral color with a alpha channel.

### `BorderColors`

One border color might not be enough for all use cases. In the past this was somewhat annoying.
Instead I think we can do something like:

```javascript
const BorderColors = {
   default: '...',
   muted: '...',
   extraMuted: '...'
}
```

#### Usage

You will still be able to pass in any color.

```jsx
<Pane borderBottom>Same as default</Pane>
<Pane borderBottom='default' /> 
<Pane borderRight='muted' /> 
<Pane borderRight='extraMuted' /> 
<Pane borderRight='red' /> 
```

**Comment by nettofarah** [comment]
Should specific border properties cancel elevation altogether? or should we issue a warning if they're both used together?

**Comment by jeroenransijn** [comment]
@nettofarah good question, I was thinking about this too. I thiought there was an exception when you deal with cards and disable border radius on a side, to attach it to something. But seems actually like no real use case for that.

Differently from this I initially had backgroundColor with some variables instead of appearances. I think we should add appearance="neutralTint1" going up to maybe 3-5. This might also set a color border if applicable. Not sure about naming just yet.

Really appreciate you take the time and give some feedback! Let's make this thing awesome 🔥

**Comment by jeroenransijn** [comment]
First pass implemented by PR #9. Had to fix the `css` in `ui-box` too.

**Comment by jeroenransijn** [comment]
This is the first pass I implemented. Note that the text isn't styled, because you should always wrap Text in a Text based component. I want to tackle documentation in a bigger swoop, so trying not to spend too much time on that right now.

<img width="1440" alt="screen shot 2017-09-04 at 5 32 51 pm" src="https://user-images.githubusercontent.com/564463/30040665-36a75722-9197-11e7-8f58-7e5cfd01b85a.png">
